components/tabs/tree_tab_node.cc

```diff
@@ -5,8 +5,81 @@
 
 #include "brave/components/tabs/public/tree_tab_node.h"
 
+#include <algorithm>
 #include <memory>
 #include <utility>
+#include <variant>
+
+#include "absl/functional/overload.h"
+#include "base/containers/adapters.h"
+#include "base/functional/bind.h"
+#include "base/types/to_address.h"
+#include "components/tabs/public/tab_interface.h"
+
+// static
+void TreeTabNode::BuildTreeTabs(TabCollection& root) {
+  auto tabs = root.GetTabsRecursive();
+
+  while (!tabs.empty()) {
+    auto* tab_interface = tabs.back();
+    tabs.pop_back();
+
+    auto* parent_collection =
+        tab_interface->GetParentCollection(GetPassKeyStatic());
+    CHECK(parent_collection) << "Tab should always have a parent collection.";
+    CHECK_NE(parent_collection->type(), TabCollection::Type::PINNED)
+        << "Pinned tabs should not be wrapped in TreeTabNode.";
+    int index = *parent_collection->GetIndexOfTab(tab_interface);
+    auto owned_tab_interface = parent_collection->MaybeRemoveTab(tab_interface);
+    CHECK_EQ(tab_interface, owned_tab_interface.get());
+
+    auto tree_node = std::make_unique<TreeTabNode>(
+        tree_tab::TreeTabNodeId::GenerateNew(), std::move(owned_tab_interface));
+    parent_collection->AddCollection(std::move(tree_node), index);
+  }
+}
+
+void TreeTabNode::FlattenTreeTabs(TabCollection& root) {
+  // Get all TreeTabNodes recursively from the root.
+  std::vector<TreeTabNode*> all_tree_nodes;
+  CollectTreeNodesRecursively(root, all_tree_nodes);
+
+  // Process tree nodes in reverse order to handle children before parents.
+  // i.e. we have all tree nodes in reverse post-order traversal.
+  for (auto* tree_node : base::Reversed(all_tree_nodes)) {
+    // Move all tabs from the tree node to its parent collection.
+
+    // * We're going to remove tab tree nodes and insert all children of it to
+    // where the tree node is.
+    std::vector<std::variant<tabs::TabInterface*, TabCollection*>> children =
+        tree_node->GetChildren();
+    auto* parent_collection = tree_node->GetParentCollection();
+    CHECK(parent_collection)
+        << "Tree node should always have a parent collection.";
+    auto index = parent_collection->GetIndexOfCollection(tree_node).value();
+
+    // * Move tabs and collection to the parent collection at the same index as
+    // the tree node.
+    for (const auto& child : base::Reversed(children)) {
+      std::visit(absl::Overload{
+                     [&](tabs::TabInterface* tab) {
+                       parent_collection->AddTab(tree_node->MaybeRemoveTab(tab),
+                                                 index);
+                     },
+                     [&](tabs::TabCollection* collection) {
+                       parent_collection->AddCollection(
+                           tree_node->MaybeRemoveCollection(collection), index);
+                     }},
+                 child);
+    }
+
+    // * Remove the tree node itself.
+    CHECK(tree_node->GetChildren().empty())
```

Comment:
This check cannot stay as is. The function for this is:

```cxx
  // Returns the direct children of this TreeTabNode as a list of variants
  // containing either TabInterface* or TabCollection*.
  std::vector<std::variant<tabs::TabInterface*, TabCollection*>> GetChildren();
```

This is a non-trivial function to be called, instantiating a `vector`. It looks like we could just use `GetChildrenStatic`, which returns a reference.

Reply:
While renaming the TreeTabNode::GetChildren(), I found that this calls base class's GetChildren() which returns const& of ChildrenVector. So it's find, isn't it?